### PR TITLE
Update tests to Intern 4

### LINKED ISSUE #617: Update tests to Intern 4

### PR DESCRIPTION
**Pull request by edhager**
Update the unit and functional tests from Intern 3 to Intern 4.

Resolves #617

### CODE REVIEW

File: intern.json [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+			"tunnel": "browserstack",
+			"capabilities+": {
+				"browserstack.debug": false
+			},
+			"environments+": [
+				{ "browserName": "internet explorer", "version": "11" },
+				{ "browserName": "edge" },
+				{ "browserName": "firefox", "platform": "WINDOWS" },
+				{ "browserName": "chrome", "platform": "WINDOWS" }
+				// Disable Safari 9 because of timing issues when running the tests.
```
Comment (agubler) [inline intern.json]: Can we remove the comments please?

File: tests/support/createTestWidget.ts [hunk 2]
```diff
@@ -1,7 +1,7 @@
-import { WidgetBase } from './../../src/WidgetBase';
-import { w } from './../../src/d';
-import { Registry } from './../../src/Registry';
-import { Constructor, VirtualDomNode, WidgetBaseInterface } from './../../src/interfaces';
+import { WidgetBase } from '../../src/WidgetBase';
```
Comment (agubler) [inline tests/support/createTestWidget.ts]: Is this something your IDE does automatically?

File: tests/unit/Registry.ts [hunk 3]
```diff
@@ -97,57 +96,54 @@ registerSuite({
 				assert.isNull(item);
 			},
 			'replaces promise with result on resolution'() {
-				let resolveFunction: any;
 				const promise: any = new Promise((resolve) => {
-					resolveFunction = resolve;
+					setTimeout(() => {
```
Comment (agubler) [inline tests/unit/Registry.ts]: Is a timeout better than resolving explicitly?

File: tests/unit/WidgetBase.ts [hunk 4]
```diff
@@ -20,1031 +20,1071 @@ function testDecorator(func?: Function) {
...
-			assert.strictEqual(result.children![0].text, 'text');
-			assert.strictEqual(result.children![1].vnodeSelector, 'span');
-			assert.strictEqual(result.children![1].properties.key, 'span');
-		},
-		'instance gets passed to VNodeProperties as bind to widget and all children'() {
+				const widget = new TestWidget();
+				const result: any = widget.__render__();
+				assert.strictEqual(result.vnodeSelector, 'div');
+				assert.lengthOf(result.children, 2);
+				assert.strictEqual(result.children![ 0 ].vnodeSelector, 'div');
```
Comment (agubler) [inline tests/unit/WidgetBase.ts]: Not sure we specifically pad indexes of arrays like this?

File: tests/unit/customElements.ts [hunk 5]
```diff
@@ -321,28 +324,29 @@ registerSuite({
...
 				let rendered = false;
 
 				const appendStub = sandbox.stub();
 
-				sandbox.stub(projector, 'ProjectorMixin', function () {
-					return {
-						append: appendStub
-					};
-				});
+				const OrigProjectorMixin = projector.ProjectorMixin;
```
Comment (agubler) [inline tests/unit/customElements.ts]: Was this causing an issue as it was?

**Comment by agubler** [comment]
Also there seem to be conflicts already.

File: tests/support/createTestWidget.ts (hunk 2 above)
Comment (edhager) [reply tests/support/createTestWidget.ts]: My IDE (IntelliJ IDEA) doesn't like paths that start with "./../".  I could turn the checks off but what is the purpose of using "./.."?

File: tests/unit/Registry.ts (hunk 3 above)
Comment (edhager) [reply tests/unit/Registry.ts]: We had a difficult time getting the registry tests to work properly.  The setTimeout calls were an experiment that now we don't need so I will remove them.

File: tests/unit/customElements.ts (hunk 5 above)
Comment (edhager) [reply tests/unit/customElements.ts]: The stubbed ProjectorMixin was reported as not being a constructor.
```
  TypeError: projector is not a constructor
      at Object.initializeElement  <src/customElements.ts:244:24>
      at Test.appender is returned as a function [as test]  <tests/unit/customElements.ts:358:21>
      at <src/lib/Test.ts:257:48>
      at <node_modules/intern/node_modules/@dojo/core/async/Task.ts:319:12>
      at handler  <node_modules/intern/node_modules/@dojo/core/async/ExtensiblePromise.ts:216:14>
      at <anonymous>
```

**Comment by edhager** [comment]
I'll merge in the updates from master and resolve the conflict.

File: tests/support/createTestWidget.ts (hunk 2 above)
Comment (agubler) [reply tests/support/createTestWidget.ts]: It's just my habit really (probably why it's all through the repo)

File: tests/unit/WidgetBase.ts (hunk 4 above)
Comment (edhager) [reply tests/unit/WidgetBase.ts]: I updated my IDE to not add the spaces and reformatted the files.

Comment (edhager) [reply tests/unit/WidgetBase.ts]: I updated my IDE settings and reformatted the files.

File: tests/functional/Drag.ts [hunk 6]
```diff
@@ -42,6 +42,9 @@ registerSuite('Drag', {
 		if (browserName === 'MicrosoftEdge') {
 			this.skip('For some reason, findById not working on Edge ATM.');
 		}
+		if (browserName === 'internet explorer') {
+			this.skip('Dragging is not working on Internet Explorer.');
```
Comment (edhager) [inline tests/functional/Drag.ts]: Note: I can load the functional test HTML page in IE11, click and drag, and the page functions as expected.  There seems to be some problem doing that through the remote interface.

**Comment by bryanforbes** [comment]
Resolved in b626059

**Pull request closed without merging**